In [1]:
i=1
while(i<=100):
    print(i)
    i += 1

1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100


In [3]:
#Print numbers from 100 to 1 using while loop
i=100
while(i>=1):
    print(i)
    i -= 1

100
99
98
97
96
95
94
93
92
91
90
89
88
87
86
85
84
83
82
81
80
79
78
77
76
75
74
73
72
71
70
69
68
67
66
65
64
63
62
61
60
59
58
57
56
55
54
53
52
51
50
49
48
47
46
45
44
43
42
41
40
39
38
37
36
35
34
33
32
31
30
29
28
27
26
25
24
23
22
21
20
19
18
17
16
15
14
13
12
11
10
9
8
7
6
5
4
3
2
1


In [4]:
i=2
while (i<=100):
    if (i%2==0):
        print(i,"even number")
    i += 1

2 even number
4 even number
6 even number
8 even number
10 even number
12 even number
14 even number
16 even number
18 even number
20 even number
22 even number
24 even number
26 even number
28 even number
30 even number
32 even number
34 even number
36 even number
38 even number
40 even number
42 even number
44 even number
46 even number
48 even number
50 even number
52 even number
54 even number
56 even number
58 even number
60 even number
62 even number
64 even number
66 even number
68 even number
70 even number
72 even number
74 even number
76 even number
78 even number
80 even number
82 even number
84 even number
86 even number
88 even number
90 even number
92 even number
94 even number
96 even number
98 even number
100 even number


In [6]:
#Calculate the sum of digits of a number.
# Initialization: Get user input and setup sum variable
number = int(input("Enter a number: "))
sum = 0

# Condition: Continue until all digits are extracted (number becomes 0)
while number > 0:
    # Extract: Get the last digit using modulus operator
    digit = number % 10

    # Accumulate: Add the extracted digit to sum
    sum += digit

    # Update/Termination step: Remove the last digit using integer division
    number //= 10

# Print final result
print("The sum of digits is:", sum)

The sum of digits is: 16


Logic:

Here is the iteration-by-iteration breakdown for the input 5832, showing the exact values of digit, number, and sum at each step of the loop.

Variables at Each Iteration
Initial Values (Before Loop Begins)

number = 5832

sum = 0

Iteration 1

digit = number % 10  →  5832 % 10 = 2

sum += digit  →  0 + 2 = 2

number //= 10  →  5832 // 10 = 583

Iteration 2

digit = number % 10  →  583 % 10 = 3

sum += digit  →  2 + 3 = 5

number //= 10  →  583 // 10 = 58

Iteration 3

digit = number % 10  →  58 % 10 = 8

sum += digit  →  5 + 8 = 13

number //= 10  →  58 // 10 = 5

Iteration 4

digit = number % 10  →  5 % 10 = 5

sum += digit  →  13 + 5 = 18

number //= 10  →  5 // 10 = 0

Loop Exit

while (number > 0) checks 0 > 0, which evaluates to False.

Loop stops. Final sum = 18.

In [7]:
#Reverse an integer using a while loop.
number = int(input("Enter a number: "))
reversed_number = 0

while number > 0:
    digit = number % 10
    reversed_number = reversed_number * 10 + digit
    number //= 10

print("The reversed number is:", reversed_number)   

The reversed number is: 7345


How the Logic Works:

digit = number % 10: Extracts the last digit of number.
reversed_number = reversed_number * 10 + digit: Shifts the current digits in reversed_number one place to the left (by multiplying by 10) and appends the extracted digit at the end.
number //= 10: Removes the last digit from number by performing integer division.



  Step-by-Step Execution Trace (Input: 12345)
Initial Values (Before Loop)

number = 12345

reversed_number = 0

   Iteration 1:
   
   digit = 12345 % 10 -> 5

   reversed_number = (0 * 10) + 5 -> 5

   number = 12345 // 10 -> 1234

   teration 2
digit = 1234 % 10 -> 4
reversed_number = (5 * 10) + 4 -> 54
number = 1234 // 10 -> 123

Iteration 3
digit = 123 % 10 -> 3
reversed_number = (54 * 10) + 3 -> 543
number = 123 // 10 -> 12

Iteration 4
digit = 12 % 10 -> 2
reversed_number = (543 * 10) + 2 -> 5432
number = 12 // 10 -> 1

Iteration 5
digit = 1 % 10 -> 1
reversed_number = (5432 * 10) + 1 -> 54321
number = 1 // 10 -> 0

Loop Exit
Condition check: number > 0 (0 > 0) is False.
The loop terminates.


In [10]:
#Count the number of digits in an integer.
number = int(input("Enter a number: "))
count = 0

while number > 0:
    number //= 10
    count += 1

print("The number of digits is:", count)


The number of digits is: 6


How the Logic Works
number //= 10: Performs integer division to remove the last digit from number.

count += 1: Increments the digit counter by 1 for every digit removed.

while number > 0: Keeps running as long as there are digits remaining in number.

   Step-by-Step Execution Trace (Input: 5832)

   Initial Values (Before Loop)number = 5832
   
   count = 0
   
   Iteration 1number = 5832 // 10  -> 583
   
   count += 1  ->  0 + 1 = 1
   
   Iteration 2:
   
   number = 583 // 10  ->  58
   
   count += 1  ->  1 + 1 = 2
   
   Iteration 3:
   
   number = 58 // 10  -> 5
   
   count += 1 -> 2 + 1 = 3
   
   Iteration 4number = 5 // 10  -> 0
   
   count += 1  -> 3 + 1 = 4

Loop Exit:

Condition check: number > 0 (0 > 0) is False.

The loop terminates.

In [11]:
# Calculate factorial using while.


# Initialization: Get user input, initialize factorial result to 1, and set counter
number = int(input("Enter a number: "))
factorial = 1
i = number

# Condition: Continue as long as counter is greater than 0
while i > 0:
    # Accumulate: Multiply factorial by current value of i
    factorial *= i
    
    # Update/Termination step: Decrement i by 1
    i -= 1

# Output final result
print(f"The factorial of {number} is:", factorial)

The factorial of 5 is: 120


Step-by-Step Execution Trace (Input: 5)

Calculating 5! = 5 x4 x3 x2 x1 = 120

Initial Values (Before Loop)

number = 5

factorial = 1

i = 5

Iteration 1:

factorial *= i  ->  1 * 5 = 5

i -= 1  ->  5 - 1 = 4

Iteration 2:

factorial *= i  ->  5 * 4 = 20

i -= 1  ->  4 - 1 = 3

Iteration 3:

factorial *= i  ->  20 * 3 = 60

i -= 1  ->  3 - 1 = 2

Iteration 4:

factorial *= i  ->  60 * 2 = 120

i -= 1  ->  2 - 1 = 1

Iteration 5:

factorial *= i -> 120 * 1 = 120

i -= 1  ->  1 - 1 = 0

Loop Exit:

Condition check: i > 0 (0 > 0) is False.

Loop terminates. Final factorial = 120.

In [ ]:
# Create a program that repeatedly asks for numbers and stops only 
# when the user enters 0. Display the sum of all previously entered numbers.

# Initialization: Variable to hold the running total, and initial user prompt
total_sum = 0
number = int(input("Enter a number (0 to stop): "))

# Condition: Keep looping as long as the user hasn't entered 0
while number != 0:
    # Accumulate: Add the entered number to the total sum
    total_sum += number
    
    # Update/Termination step: Ask for the next number
    number = int(input("Enter a number (0 to stop): "))

# Output final result
print("The sum of all entered numbers is:", total_sum)

The sum of all entered numbers is: 153


Step-by-Step Execution Trace:

Using sample input numbers in order: 10, 5, 7, 0.

Initial State (Before Loop)

total_sum = 0

User inputs 10 -> number = 10

Iteration 1:

Condition Check: number != 0 (10 != 0) is True

total_sum += number -> 0 + 10 = 10

User Prompt: User inputs 5 $-> number = 5

Iteration 2:

Condition Check: number != 0 (5 != 0) is True

total_sum += number -> 10 + 5 = 15

User Prompt: User inputs 7 -> number = 7

Iteration 3:

Condition Check: number != 0 (7 != 0) is True

total_sum += number -> 15 + 7 = 22

User Prompt: User inputs 0 -> number = 0

Loop Exit:

Condition Check: number != 0 (0 != 0) is False

The loop terminates without adding 0 to the total.


In [13]:
#Create a password checker that keeps asking for the password until the correct password is entered.
correct_password = "Python123"
password = input("Enter the password: ")

while password != correct_password:
    print("Incorrect password. Please try again.")
    password = input("Enter the password: ")

print("Password accepted.")

Incorrect password. Please try again.
Incorrect password. Please try again.
Password accepted.


In [14]:
#Create a guessing game where the secret number is predefined and the user keeps guessing until correct.
secret_number = 7
guess = int(input("Guess the number (1-10): "))

while guess != secret_number:
    print("Wrong guess. Try again.")
    guess = int(input("Guess the number (1-10): "))

print("Congratulations! You guessed the number.")

Wrong guess. Try again.
Wrong guess. Try again.
Wrong guess. Try again.
Wrong guess. Try again.
Congratulations! You guessed the number.


In [15]:
#Build a menu-driven calculator
print("Menu:")
print("1. Add")
print("2. Subtract")
print("3. Multiply")
print("4. Divide")
print("5. Exit")

choice = int(input("Enter your choice (1-5): "))

while choice != 5:
    if choice == 1:
        # Add logic here
        pass
    elif choice == 2:
        # Subtract logic here
        pass
    elif choice == 3:
        # Multiply logic here
        pass
    elif choice == 4:
        # Divide logic here
        pass
    else:
        print("Invalid choice. Please try again.")

    print("\nMenu:")
    print("1. Add")
    print("2. Subtract")
    print("3. Multiply")
    print("4. Divide")
    print("5. Exit")
    choice = int(input("Enter your choice (1-5): "))

print("Calculator exited.")


Menu:
1. Add
2. Subtract
3. Multiply
4. Divide
5. Exit

Menu:
1. Add
2. Subtract
3. Multiply
4. Divide
5. Exit

Menu:
1. Add
2. Subtract
3. Multiply
4. Divide
5. Exit
Calculator exited.


In [16]:
# Initialization: Variable to track user choice
choice = 0

# Condition: Continue looping as long as the choice is not 5 (Exit)
while choice != 5:
    # Display the menu options
    print("\n--- Calculator Menu ---")
    print("1. Add")
    print("2. Subtract")
    print("3. Multiply")
    print("4. Divide")
    print("5. Exit")
    
    # Update/Termination step (1): Get user choice
    choice = int(input("Enter your choice (1-5): "))
    
    # Check if user wants to exit
    if choice == 5:
        print("Exiting calculator. Goodbye!")
    elif choice in (1, 2, 3, 4):
        # Prompt for numbers only when a valid arithmetic operation is selected
        num1 = float(input("Enter first number: "))
        num2 = float(input("Enter second number: "))
        
        # Perform operation based on choice
        if choice == 1:
            print(f"Result: {num1} + {num2} = {num1 + num2}")
        elif choice == 2:
            print(f"Result: {num1} - {num2} = {num1 - num2}")
        elif choice == 3:
            print(f"Result: {num1} * {num2} = {num1 * num2}")
        elif choice == 4:
            # Check for division by zero
            if num2 != 0:
                print(f"Result: {num1} / {num2} = {num1 / num2}")
            else:
                print("Error: Division by zero is not allowed.")
    else:
        print("Invalid choice! Please select a valid option from 1 to 5.")


--- Calculator Menu ---
1. Add
2. Subtract
3. Multiply
4. Divide
5. Exit
Result: 4.0 * 7.0 = 28.0

--- Calculator Menu ---
1. Add
2. Subtract
3. Multiply
4. Divide
5. Exit
Exiting calculator. Goodbye!


In [17]:
# Initialization: Initial balance and choice variable
balance = 1000.0  # Starting balance
choice = 0

# Condition: Continue running as long as the user does not choose option 4 (Exit)
while choice != 4:
    # Display the ATM Menu
    print("\n--- ATM Menu ---")
    print("1. Check Balance")
    print("2. Deposit Money")
    print("3. Withdraw Money")
    print("4. Exit")
    
    # Update/Termination step (1): Get user choice
    choice = int(input("Choose an option (1-4): "))
    
    # Process user choice
    if choice == 1:
        print(f"Your current balance is: ${balance:.2f}")
        
    elif choice == 2:
        amount = float(input("Enter amount to deposit: $"))
        if amount > 0:
            balance += amount
            print(f"Successfully deposited ${amount:.2f}. New balance: ${balance:.2f}")
        else:
            print("Invalid deposit amount!")
            
    elif choice == 3:
        amount = float(input("Enter amount to withdraw: $"))
        if amount > balance:
            print("Insufficient funds! Transaction cancelled.")
        elif amount <= 0:
            print("Invalid withdrawal amount!")
        else:
            balance -= amount
            print(f"Successfully withdrew ${amount:.2f}. New balance: ${balance:.2f}")
            
    elif choice == 4:
        print("Thank you for using the ATM. Goodbye!")
        
    else:
        print("Invalid selection! Please choose an option between 1 and 4.")


--- ATM Menu ---
1. Check Balance
2. Deposit Money
3. Withdraw Money
4. Exit
Your current balance is: $1000.00

--- ATM Menu ---
1. Check Balance
2. Deposit Money
3. Withdraw Money
4. Exit
Successfully withdrew $5.00. New balance: $995.00

--- ATM Menu ---
1. Check Balance
2. Deposit Money
3. Withdraw Money
4. Exit
Thank you for using the ATM. Goodbye!
